nvflare/app_common/executors/launcher_executor.py

```diff
@@ -95,51 +101,38 @@ def __init__(
         self._launcher_finish_status = None
         self._launcher_finish_time = None
         self._last_result_transfer_timeout = last_result_transfer_timeout
+        self._received_result = False
         self._job_end = False
 
         self._thread_pool_executor = ThreadPoolExecutor(max_workers=workers, thread_name_prefix=self.__class__.__name__)
 
-        self.pipe_handler: Optional[PipeHandler] = None
-        self._pipe: Optional[Pipe] = None
-        self._pipe_id = pipe_id
-        self._pipe_name = None
-        self._read_interval = read_interval
-        self._heartbeat_interval = heartbeat_interval
-        self._heartbeat_timeout = heartbeat_timeout
-        self._result_timeout = result_timeout
         self._result_poll_interval = result_poll_interval
-        self._peer_read_timeout = peer_read_timeout
 
         # flags to indicate whether the launcher side will send back trained model and/or metrics
         self._train_with_evaluation = train_with_evaluation
         self._train_task_name = train_task_name
         self._evaluate_task_name = evaluate_task_name
         self._submit_model_task_name = submit_model_task_name
-        self._result_fl_model = None
-        self._result_metrics = None
 
-        self._from_nvflare_converter_id = from_nvflare_converter_id
-        self._from_nvflare_converter: Optional[ParamsConverter] = None
-        self._to_nvflare_converter_id = to_nvflare_converter_id
-        self._to_nvflare_converter: Optional[ParamsConverter] = None
+        self._error_happened = threading.Event()
+        self._check_last_result_transfer_thread = None
 
     def initialize(self, fl_ctx: FLContext) -> None:
         self._init_launcher(fl_ctx)
-        self._init_converter(fl_ctx)
-        self._init_pipe(fl_ctx)
+        self.init_pipe(fl_ctx)
 
     def handle_event(self, event_type: str, fl_ctx: FLContext) -> None:
         if event_type == EventType.START_RUN:
             self.initialize(fl_ctx)
+            super().handle_event(event_type, fl_ctx)
         elif event_type == EventType.END_RUN:
             if self.launcher is None:
                 raise RuntimeError("Launcher is None.")
             self._job_end = True
+            self._error_happened.clear()
             self.launcher.finalize(fl_ctx)
             self.log_info(fl_ctx, "END_RUN event received - telling external to stop")
-            if self.pipe_handler is not None:
-                self.pipe_handler.notify_end("END_RUN event received")
-                self.pipe_handler.stop(close_pipe=True)
+            super().handle_event(event_type, fl_ctx)
 
     def execute(self, task_name: str, shareable: Shareable, fl_ctx: FLContext, abort_signal: Signal) -> Shareable:
```

Comment:
This code should be re-organized for better readability. This executor has the following 3 major steps:
- Launch the 3rd-party program
- Execute the task
- End the 3rd-party program

Reply:
re-organized as suggested.